[Index Management] Implement index overview cards

yuliacech opened the pull request:
## Summary

This PR implements a new design for the Overview tab on the index details page. There are 4 cards, but only 3 can be displayed at one time: Storage, Status and Aliases/Data Stream (a backing index of the data stream can't have any aliases). On Serverless, there is no Storage and Status cards, so the whole section is either empty or 1 card is displayed for Aliases or Data Stream. 

### Screenshots
#### (Stateful) Index without aliases and data stream
<img width="1335" alt="Screenshot 2023-10-27 at 17 25 56" src="https://github.com/elastic/kibana/assets/6585477/38e4f96b-d612-42a7-9d2d-431d3ff4cd7b">


#### (Stateful) Index with aliases
<img width="1332" alt="Screenshot 2023-10-27 at 17 27 05" src="https://github.com/elastic/kibana/assets/6585477/4a3b03ec-8da8-47dd-b343-ba83ddfe5efd">


Aliases flyout 
<img width="645" alt="Screenshot 2023-10-27 at 17 27 39" src="https://github.com/elastic/kibana/assets/6585477/4f4770dd-ce3e-4bf7-9e50-d2a58645ea0b">


#### (Stateful) Backing index of a data stream
<img width="1340" alt="Screenshot 2023-10-27 at 17 28 16" src="https://github.com/elastic/kibana/assets/6585477/434e3aec-06bb-479f-be2e-37d3d2b74951">



#### (Serverless) Index without cards
<img width="1264" alt="Screenshot 2023-10-09 at 16 16 04" src="https://github.com/elastic/kibana/assets/6585477/5b40f61e-0e14-47ee-9ac5-a91ffe14775b">

#### (Serverless) Index with aliases
<img width="1260" alt="Screenshot 2023-10-09 at 16 16 34" src="https://github.com/elastic/kibana/assets/6585477/193b6a6f-8caa-4e77-9ec8-31c133584e75">

#### (Serverless) Backing index of a data stream
<img width="1598" alt="Screenshot 2023-10-24 at 17 29 53" src="https://github.com/elastic/kibana/assets/6585477/602c09fb-1b85-4d12-88b2-bf9eb78639b4">

### Screen recordings
#### (Stateful) Index with aliases

https://github.com/elastic/kibana/assets/6585477/c44dc6a7-dc2b-4297-88ae-1ecc22f828da



#### (Stateful) Backing index of a data stream

https://github.com/elastic/kibana/assets/6585477/5206e427-b373-4ec8-a2ca-253b125afaee





### How to test
- Start either a stateful or serverless ES and Kibana: `yarn es snapshot` and `yarn start` or `yarn es serverless --ssl` and `yarn serverless-es --ssl`
- Use following commands in Dev Tools Console to create an index and add some aliases
```
PUT /test_index

POST _aliases
{
  "actions": [
    {
      "add": {
        "index": "test_index",
        "alias": "test_alias_1"
      }
    }
  ]
}
```
- Use following commdans in Dev Tools to create a data stream
```
PUT _index_template/test-index-template
{
  "index_patterns": ["test-data-stream*"],
  "data_stream": { },
  "priority": 500
}

PUT _data_stream/test-data-stream
```

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

sabarasaba commented on x-pack/plugins/index_management/public/application/sections/home/index_list/details_page/details_page_overview/data_stream_details.tsx:
```diff
@@ -0,0 +1,154 @@
+      <EuiFlexItem grow={false}>
+        <EuiText
+          css={css`
+            font-size: ${euiThemeVars.euiFontSizeL};
```
wouldnt be the same (and perhaps cleaner) to just use the `size="L"` from `EuiText` instead? 🤔

yuliacech replied:
There is no `"L"` for text sizes for some reason, only `["xs", "s", "m", "relative"]` 🤔

sabarasaba replied:
that is soooo strange! whyyyy! Your changes lgtm in the mean time tho

sabarasaba commented on x-pack/plugins/index_management/public/application/sections/home/index_list/details_page/details_page_overview/details_page_overview.tsx:
```diff
@@ -65,99 +73,28 @@ export const DetailsPageOverview: React.FunctionComponent<Props> = ({ indexDetai
-                </EuiFlexItem>
-              </EuiFlexGroup>
-            </EuiPanel>
+      <EuiFlexGrid columns={isLarge ? 3 : 2}>
```
If we ever run into a scenario in which we only have one column (for example in serverless), is there any chance we could show some other data so that it wouldnt end up being that empty?

Otherwise I think might look better if we just use one column and have the only card to be expanded:

![Screenshot 2023-10-27 at 13 22 52](https://github.com/elastic/kibana/assets/1191206/595f3414-95a8-47d8-aa74-33ccb90f32ed)

yuliacech replied:
On serverless there is no other info to display, so it might be either nothing, or only aliases card, or only data stream card. I personally would prefer the card to stay as a 1/3 of the width even when there is only 1 card. But let me get some feedback from the design team on that.

sabarasaba replied:
Yeah I'll leave it up design on what to do 👍🏼

yuliacech replied:
Sent your screenshot to the design team and will follow up in a separate PR if needed

sabarasaba commented on x-pack/plugins/index_management/public/application/sections/home/index_list/details_page/details_page_overview/status_details.tsx:
```diff
@@ -0,0 +1,105 @@
+    return null;
+  }
+  let healthBadge: ReactNode;
+  if (health === 'green' || health === 'GREEN') {
```
would this be a scenario we might run into (discrepancies in capitalization)? Otherwise we could try to normalize it like: `health.toLowerCase() === 'green'`.. etc

Bit nit picky, but we could also avoid having to assign Badges to a variable if we use an object lookup to define the badge:

```typescript

const healthBadge = {
  red: {
    color: 'danger',
    label: 'Critical'
  },
  ...
};

const healthStatus = healthBadge[health];
const healthBadge = (
  <EuiBadge color={healthStatus.color}>
      {healthStatus.label}
  </EuiBadge>
);

```

yuliacech replied:
That's a great suggestion, @sabarasaba, thanks! I'll add it to the next commit.

sabarasaba commented on x-pack/plugins/index_management/public/application/sections/home/index_list/details_page/details_page_overview/status_details.tsx:
```diff
@@ -0,0 +1,105 @@
+  if (!config.enableIndexStats) {
+    return null;
+  }
+  let healthBadge: ReactNode;
```
I know its a bit out of scope for this PR so feel free to ignore, but dont we have more information on why the health status might be critical or warning? Feels like we are telling things to the user but they might not know what caused it or how to fix it

yuliacech replied:
Good point, opened https://github.com/elastic/kibana/issues/170040 to track this

sabarasaba commented on x-pack/plugins/index_management/public/application/sections/home/index_list/details_page/details_page_overview/aliases_details.tsx:
```diff
@@ -0,0 +1,113 @@
+  if (!Array.isArray(aliases)) {
+    return null;
+  }
+  const aliasesBadges = aliases.slice(0, 3).map((alias) => <EuiBadge>{alias}</EuiBadge>);
```
Should we truncate the text for each alias? if some of them are long it might look a bit off in the ui

![Screenshot 2023-10-27 at 13 23 54](https://github.com/elastic/kibana/assets/1191206/80ce6f0b-9eb6-4dab-a2fb-e09abf998202)

yuliacech replied:
Great point, I think I could set a sensible max width value, something like 250px

sabarasaba commented on x-pack/plugins/index_management/public/application/sections/home/index_list/details_page/details_page_overview/status_details.tsx:
```diff
@@ -0,0 +1,105 @@
+          left: (
+            <EuiText
+              color={status === 'close' ? 'danger' : 'success'}
+              css={css`
```
wouldnt be the same (and perhaps cleaner) to just use the size="L" from EuiText instead? 🤔

sabarasaba reviewed (commented):
Thanks for working on this @yuliacech! I went through the changes and left some comments, let me know what you think!

There's also a few more things I thought about pointing out, at least to have a conversation:

* At some point in a specific breakdown point at around 800px width the header CTAs go bananas and looks a bit off:

![Screenshot 2023-10-27 at 13 16 07](https://github.com/elastic/kibana/assets/1191206/f272765b-0ab9-4834-bd95-fdbd92dc3a10)

* For smaller screen sizes I dont think we tend to pay too much attention to that (wronfully, but still) since I dont think most of our users use kibana that way. So not sure how much attention to detail we should give to small screen devices. 

* When we show 3 cards, if we size down the browser width at some point we will end up with 2 cards at the top side and one below. I think we could change the cards container css property for the template-columns to be of just one row `grid-template-columns: repeat(1, 1fr);` and with that we could get rid of the empty space:


https://github.com/elastic/kibana/assets/1191206/14b40f48-71c6-4b74-b3d0-728c53e9e110

Let me know what you think!

yuliacech commented on x-pack/plugins/index_management/public/application/sections/home/index_list/details_page/details_page_overview/aliases_details.tsx:
```diff
@@ -0,0 +1,113 @@
+          </EuiFlyoutHeader>
+          <EuiFlyoutBody>
+            <EuiListGroup>
+              {aliases.map((alias) => (
```
I didn't notice that, great catch, I'll make the list to stretch for the whole width of the flyout and wrap the text if it's too long.

yuliacech commented:
Thank you so much for such a detailed code review, @sabarasaba! 
I think I addressed all your comments in the code, I added changes in the latest commits, also fixed the issue with the buttons header and updated the cards layout to be 3 on a larger screen and 1 on a smaller screen as you suggested. I updated the screenshots in the PR description that should reflect the latest updates. Could you please have another look?

sabarasaba commented on x-pack/plugins/index_management/public/application/sections/home/index_list/details_page/details_page_overview/aliases_details.tsx:
```diff
@@ -0,0 +1,123 @@
+  if (!Array.isArray(aliases)) {
+    return null;
+  }
+  const aliasesBadges = aliases.slice(0, 3).map((alias) => (
```
```suggestion
  const aliasesBadges = aliases.slice(0, MAX_VISIBLE_ALIASES).map((alias) => (
```

sabarasaba reviewed (approved):
Latest lgtm, thanks for addressing the changes @yuliacech!
